src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADBatchTaskRunner.java

```diff
@@ -188,47 +394,181 @@ public ADBatchTaskRunner(
-                            node.getId(),
-                            adTask.getDetectorId()
-                        );
-                    transportService
-                        .sendRequest(
-                            node,
-                            ADBatchTaskRemoteExecutionAction.NAME,
-                            new ADBatchAnomalyResultRequest(adTask),
-                            option,
-                            new ActionListenerResponseHandler<>(delegatedListener, ADBatchAnomalyResultResponse::new)
-                        );
+                    adTaskCacheManager.removeEntity(adTask.getDetectorId(), adTask.getEntity().get(0).getValue());
+                    logger.warn("Entity task failed, task id: {}", adTask.getTaskId());
                 }
-            }, e -> delegatedListener.onFailure(e))), e -> delegatedListener.onFailure(e)));
+                // Sleep some time before polling next entity task.
+                waitBeforeNextEntity(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILIS);
+                adTaskManager.entityTaskDone(adTask, e, transportService);
+                startNewEntityTaskLane(adTask, transportService);
+            }
+        });
+
+        ThreadedActionListener threadedActionListener = new ThreadedActionListener<>(
+            logger,
+            threadPool,
+            AD_BATCH_TASK_THREAD_POOL_NAME,
+            actionListener,
+            false
+        );
+        return threadedActionListener;
+    }
+
+    private void waitBeforeNextEntity(long time) {
+        try {
+            Thread.sleep(time);
+        } catch (InterruptedException interruptedException) {
+            logger.warn("Exception while waiting", interruptedException);
+        }
+    }
+
+    private void executeSingleEntityTask(
+        ADTask adTask,
+        TransportService transportService,
+        ActionListener<ADBatchAnomalyResultResponse> delegatedListener
+    ) {
+        dispatchTask(adTask, ActionListener.wrap(node -> {
+            if (clusterService.localNode().getId().equals(node.getId())) {
+                // Execute batch task locally
+                startADBatchTask(adTask, false, transportService, delegatedListener);
+            } else {
+                // Execute batch task remotely
+                transportService
+                    .sendRequest(
+                        node,
+                        ADBatchTaskRemoteExecutionAction.NAME,
+                        new ADBatchAnomalyResultRequest(adTask),
+                        option,
+                        // TODO: check if still need to add true/false in startADBatchTask
+                        new ActionListenerResponseHandler<>(delegatedListener, ADBatchAnomalyResultResponse::new)
+                    );
+            }
+        }, e -> delegatedListener.onFailure(e)));
+    }
+
+    // start new entity task lane
+    private void startNewEntityTaskLane(ADTask adTask, TransportService transportService) {
```

Comment:
Is one lane for each entity task?  After finishing one entity task, you start a new lane by calling this function?

Reply:
Reply on new PR https://github.com/opensearch-project/anomaly-detection/pull/104/files#r660170298